alvr/client_core/src/platform/android/decoder.rs

```diff
@@ -144,6 +144,43 @@ impl Drop for VideoDecoderSource {
     }
 }
 
+// Configure & start a MediaCodec.
+fn decoder_configure_and_start(
+    decoder: MediaCodec,
+    format: &MediaFormat,
+    image_reader: &ImageReader,
+) -> Result<MediaCodec> {
+    decoder.configure(
+        &format,
+        Some(&image_reader.window().unwrap()),
+        MediaCodecDirection::Decoder,
+    )?;
+    decoder.start()?;
+    Ok(decoder)
+}
+
+// Creates a hardware (or assumed hardware) MediaCodec and then configure and start it.
+fn decoder_try_hardware(
+    mime: &str,
+    format: &MediaFormat,
+    image_reader: &ImageReader,
+) -> Result<MediaCodec> {
+    let tmp = MediaCodec::from_decoder_type(&mime)
+        .ok_or(anyhow!("unable to find decoder for mime type: {}", &mime))?;
+    decoder_configure_and_start(tmp, &format, &image_reader)
+}
+
+// Creates a software MediaCodec and then configure and start it.
+fn decoder_try_software(
+    codec_name: &str,
+    format: &MediaFormat,
+    image_reader: &ImageReader,
+) -> Result<MediaCodec> {
+    let tmp = MediaCodec::from_codec_name(&codec_name)
+        .ok_or(anyhow!("no such codec: {}", &codec_name))?;
+    decoder_configure_and_start(tmp, &format, &image_reader)
+}
```

Comment:
I believe you can unify these functions into one. Pass parameters for "codec" and "is_software", specify mime/codec_name as inline strings.

Reply:
Thanks :)